### PR TITLE
Format missing module errors in upgrade errors

### PR DESCRIPTION
**Pull request by jordanjennings-mysten**
## Description 

Missing modules errors do not have a file to point to since their file has been removed during upgrade or otherwise unknown since the declaration is gone from its previous file if the file still exists. Align missing modules errors with the behavior of other errors by using Move.toml as the reference point for these errors and pointing to the "package" instead.

## Test plan 

snapshots

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

File: crates/sui/src/upgrade_compatibility.rs [hunk 1]
```diff
@@ -813,6 +821,47 @@ fn module_compatibility_error_diag(
...
+        .unwrap_or_default();
+
+    let end = chars[start..]
+        .iter()
+        .enumerate()
+        .find(|(_, c)| c.is_whitespace())
+        .map(|(i, _)| i)
+        .unwrap_or_default();
+
+    let loc = Loc::new(hash, start as ByteIndex, end as ByteIndex);
```
Comment (amnn) [inline crates/sui/src/upgrade_compatibility.rs:850]: Using character offsets will not always map to byte offsets, unfortunately. In Rust, a `char` corresponds to a single Unicode Scalar Value, which may be encoded in between 1 to 4 UTF8 bytes in the `&str`, so you are better off working with the `&str` directly, and calling [`find`](https://doc.rust-lang.org/std/primitive.str.html#method.find) on it.

In terms of what to look for, I would try and find the `[package]` section and slice out some number of lines after it to have a good chance of finding a name.

File: crates/sui/src/upgrade_compatibility.rs [hunk 2]
```diff
@@ -813,6 +821,47 @@ fn module_compatibility_error_diag(
     Ok(diags)
 }
 
+const PACKAGE_TABLE: &str = "[package]";
```
Comment (amnn) [inline crates/sui/src/upgrade_compatibility.rs:824]: nit: you can move this inside the function body if this is the only place this constant is used (or to the top of the file if you plan on using it in more places).

File: crates/sui/src/upgrade_compatibility.rs [hunk 3]
```diff
@@ -813,6 +821,47 @@ fn module_compatibility_error_diag(
...
+    // default to the end of the package table definition
+    let mut end = start + PACKAGE_TABLE.len();
+
+    // get the third newline after the start of the package table declaration if it exists
+    toml_str
+        .match_indices('\n')
+        .filter(|(idx, _)| *idx > start)
+        .take(3)
+        .last()
+        .map(|(idx, _)| end = idx);
```
Comment (amnn) [inline crates/sui/src/upgrade_compatibility.rs:848]: Slightly more idiomatic:

```suggestion
    // Include up to three newlines after the start of the package table declaration, or
    // default to the end of the `[package]` key.
    let end = toml_str[start..]
        .match_indices('\n')
        .take(3)
        .last()
        .unwrap_or(start + PACKAGE_TABLE.len());
```

File: crates/sui/src/upgrade_compatibility.rs [hunk 4]
```diff
@@ -813,6 +821,47 @@ fn module_compatibility_error_diag(
...
+    let mut diags = Diagnostics::new();
+
+    // read Move.toml to get the hash and first line start and end
+    let toml_path = package_path.join("Move.toml");
+    let toml_str = fs::read_to_string(&toml_path).context("Unable to read Move.toml")?;
+    let hash = FileHash::new(&toml_str);
+
+    let start: usize = toml_str
+        .find(PACKAGE_TABLE)
+        .context("Malformed Move.toml")?;
```
Comment (amnn) [inline crates/sui/src/upgrade_compatibility.rs:838]: It looks like if we fail to read the `Move.toml` we may end up hiding all the other diagnostics. Do you think it would be possible for us to always produce some kind of `Diagnostics` value, even if we failed to  read the `Move.toml` to avoid that?

**Review by amnn — COMMENTED** [review]
I'm pretty happy with this change, modulo the question about trying to avoid returning an `Err(...)` instead of the diagnostic. It would be good to get @tnowacki's take on things, because he's more the expert in this area than me, but after that, I think this is good to go!

File: crates/sui/src/upgrade_compatibility.rs (hunk 4 above)
Comment (jordanjennings-mysten) [reply crates/sui/src/upgrade_compatibility.rs:838]: I was thinking this more as a "critical" error however I think it could be good to just let this return a diagnostic for byte ranges `0,0` as a fallback. Some errors I'd guess will just fall in that "critical" category and should be errored out but we could potentially discuss follow up work to prevent errors from stopping execution by converting `Errors -> Diags`.   Or alternatively just maintain a list of errors and print them last / concatenate them into a single error that is thrown last.

File: crates/sui/src/upgrade_compatibility.rs [hunk 5]
```diff
@@ -589,23 +593,11 @@ fn compare_packages(
...
-                    file_contents,
-                ),
-            );
-
-            file_set.insert(&compiled_unit_with_source.source_path);
-        }
+        insert_into_files(
+            &mut file_set,
+            &mut files,
+            &compiled_unit_with_source.source_path,
```
Comment (tnowacki) [inline crates/sui/src/upgrade_compatibility.rs:599]: Under what conditions is this file not in the fileset already? 

Would it make more sense to just up front make a file set from all of the source paths + the toml? I'm not sure we will notice a perf difference versus what we are doing here ,and that feels a bit easier to follow than what is going on here.

File: crates/sui/src/upgrade_compatibility.rs [hunk 6]
```diff
@@ -499,12 +502,13 @@ pub(crate) async fn check_compatibility(
     let policy =
         UpgradePolicy::try_from(upgrade_policy).map_err(|_| anyhow!("Invalid upgrade policy"))?;
 
-    compare_packages(existing_modules, new_package, policy)
+    compare_packages(existing_modules, new_package, package_path, policy)
 }
 
 fn compare_packages(
     existing_modules: Vec<CompiledModule>,
     new_package: CompiledPackage,
```
Comment (tnowacki) [inline crates/sui/src/upgrade_compatibility.rs:510]: We already had to parse the files for `CompiledPackage`. 
It would probably be best as a separate PR, but we should just save the file map in there so we don't have to reread the files in here.

File: crates/sui/src/upgrade_compatibility.rs [hunk 7]
```diff
@@ -813,6 +821,43 @@ fn module_compatibility_error_diag(
...
+    let mut diags = Diagnostics::new();
+
+    // read Move.toml to get the hash and first line start and end
+    let toml_path = package_path.join("Move.toml");
+    let toml_str = fs::read_to_string(&toml_path).context("Unable to read Move.toml")?;
+    let hash = FileHash::new(&toml_str);
+
+    let start: usize = toml_str.find(PACKAGE_TABLE).unwrap_or_default();
+    // default to the end of the package table definition
+    // get the third newline after the start of the package table declaration if it exists
```
Comment (tnowacki) [inline crates/sui/src/upgrade_compatibility.rs:838]: This feels a tad strange to me. I think it might make more sense point to the package identifier. e.g. 
```
[package]
name = "clowntown"
//     ^^^^^^^^^^^ point here
```

File: crates/sui/src/upgrade_compatibility.rs [hunk 8]
```diff
@@ -813,6 +821,43 @@ fn module_compatibility_error_diag(
     Ok(diags)
 }
 
+fn missing_module_diag(
+    module_name: &Identifier,
+    package_path: &PathBuf,
+) -> Result<Diagnostics, Error> {
+    const PACKAGE_TABLE: &str = "[package]";
```
Comment (tnowacki) [inline crates/sui/src/upgrade_compatibility.rs:828]: @tzakian is this string stored somewhere already in the toml parsing?

File: crates/sui/src/upgrade_compatibility.rs [hunk 9]
```diff
@@ -813,6 +821,43 @@ fn module_compatibility_error_diag(
...
+    let mut diags = Diagnostics::new();
+
+    // read Move.toml to get the hash and first line start and end
+    let toml_path = package_path.join("Move.toml");
+    let toml_str = fs::read_to_string(&toml_path).context("Unable to read Move.toml")?;
```
Comment (tnowacki) [inline crates/sui/src/upgrade_compatibility.rs:833]: Didn't we parse the contents already for the FilesSourceText?

File: crates/sui/src/upgrade_compatibility.rs (hunk 7 above)
Comment (jordanjennings-mysten) [reply crates/sui/src/upgrade_compatibility.rs:838]: I considered this but didn't want to go down the path of attempting to make a sourcemap for the toml file. That could have a lot of uses if we had it including this PR. for now the pattern is just to point to the [package] + a few lines after it.

File: crates/sui/src/unit_tests/snapshots/sui__upgrade_compatibility__upgrade_compatibility_tests__addresses_first.snap [hunk 10]
```diff
@@ -0,0 +1,13 @@
+---
+source: crates/sui/src/unit_tests/upgrade_compatibility_tests.rs
+expression: output
+---
+error[Compatibility E01007]: module missing
+  ┌─ /Users/jordanjennings/code/sui/crates/sui/src/unit_tests/fixtures/upgrade_errors/missing_module_toml/addresses_first/Move.toml:4:1
+  │  
+4 │ ╭ [package]
+5 │ │ name = "upgrades"
```
[the remark this answers is not in the document]
Comment (jordanjennings-mysten) [reply crates/sui/src/unit_tests/snapshots/sui__upgrade_compatibility__upgrade_compatibility_tests__addresses_first.snap:9]: [package] + a few lines after

File: crates/sui/src/unit_tests/snapshots/sui__upgrade_compatibility__upgrade_compatibility_tests__declarations_missing.snap [hunk 11]
```diff
@@ -38,6 +38,16 @@ error[Compatibility E01001]: missing public declaration
   = structs are part of a module's public interface and cannot be removed or changed during an upgrade.
   = add missing struct 'StructToBeRemoved' back to the module 'struct_'.
 
-The following modules are missing from the new package: 'missing_module'
+error[Compatibility E01007]: module missing
+  ┌─ /Users/jordanjennings/code/sui/crates/sui/src/unit_tests/fixtures/upgrade_errors/declaration_errors_v2/Move.toml:1:1
+  │  
+1 │ ╭ [package]
+2 │ │ name = "upgrades"
+3 │ │ edition = "2024.beta" # edition = "legacy" to use legacy (pre-2024) Move
```
[the remark this answers is not in the document]
Comment (jordanjennings-mysten) [reply crates/sui/src/unit_tests/snapshots/sui__upgrade_compatibility__upgrade_compatibility_tests__declarations_missing.snap:46]: the two options I considered were:
- write a simple pattern matching implementation to show this package table which is done here
- create a toml parser to generate Locs for '[package]' and 'name =' etc

The second would certainly be better and multiple uses but I think it would be good to get some buy in on the larger piece of work if we find this important and the toml sourcemap useful elsewhere.

File: crates/sui/src/unit_tests/snapshots/sui__upgrade_compatibility__upgrade_compatibility_tests__emoji.snap [hunk 12]
```diff
@@ -0,0 +1,15 @@
+---
+source: crates/sui/src/unit_tests/upgrade_compatibility_tests.rs
+expression: output
+---
+error[Compatibility E01007]: module missing
+  ┌─ /Users/jordanjennings/code/sui/crates/sui/src/unit_tests/fixtures/upgrade_errors/missing_module_toml/emoji/Move.toml:1:2
+  │  
+1 │   😀[package]😀
```
[the remark this answers is not in the document]
Comment (jordanjennings-mysten) [reply crates/sui/src/unit_tests/snapshots/sui__upgrade_compatibility__upgrade_compatibility_tests__emoji.snap:8]: Since emojis are 4 bytes this helps ensure no issues with reading bytes vs reading characters in the  `missing_module_diag` I'll leave a comment so it's more clear why I think this test is useful.

File: crates/sui/src/upgrade_compatibility.rs (hunk 5 above)
Comment (jordanjennings-mysten) [reply crates/sui/src/upgrade_compatibility.rs]: done

File: crates/sui/src/upgrade_compatibility.rs (hunk 6 above)
Comment (jordanjennings-mysten) [reply crates/sui/src/upgrade_compatibility.rs]: done

File: crates/sui/src/upgrade_compatibility.rs (hunk 8 above)
Comment (jordanjennings-mysten) [reply crates/sui/src/upgrade_compatibility.rs:814]: doesnt appear to be in the file_map, I'll just read it for now we might have this later with some work from @tzakian

File: crates/sui/src/upgrade_compatibility.rs (hunk 9 above)
Comment (jordanjennings-mysten) [reply crates/sui/src/upgrade_compatibility.rs]: fixed

File: crates/sui/src/upgrade_compatibility.rs [hunk 13]
```diff
@@ -590,53 +602,33 @@ fn compare_packages(
...
             compiled_unit_with_source,
             &lookup[&name],
         )?);
     }
 
+    new_package.package.file_map.add(
+        FileHash::new(&move_toml_contents),
+        FileName::from(move_toml_path.to_string_lossy()),
+        move_toml_contents,
+    );
```
Comment (tnowacki) [inline crates/sui/src/upgrade_compatibility.rs:616]: nit, move this next to the reading of the toml

**Review by tnowacki — APPROVED** [review]
